diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 9ad1a821464357ccbffdca9a3e599778d9887954..d713178d9c66c81484c7c15f34c26e6e6671a995 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -569,7 +569,8 @@ static void dequeue_pushable_dl_task(struct rq *rq, struct task_struct *p)
 	if (RB_EMPTY_NODE(&p->pushable_dl_tasks))
 		return;
 
-	leftmost = rb_erase_cached(&p->pushable_dl_tasks, root);
+	rb_erase_cached(&p->pushable_dl_tasks, root);
+	leftmost = rb_first_cached(root);
 	if (leftmost)
 		dl_rq->earliest_dl.next = __node_2_pdl(leftmost)->dl.deadline;
 
