diff --git a/arch/arm64/configs/vendor/xiaomi/magictime-common.config b/arch/arm64/configs/vendor/xiaomi/magictime-common.config
index f84dcd846ebc..0a9dfef833b6 100644
--- a/arch/arm64/configs/vendor/xiaomi/magictime-common.config
+++ b/arch/arm64/configs/vendor/xiaomi/magictime-common.config
@@ -62,8 +62,6 @@ CONFIG_ARM64_VHE=y
 # BLK
 CONFIG_BLK_WBT=n
 
-# KernelSU
-CONFIG_KSU=y
 
 # USB
 CONFIG_USB_MASS_STORAGE=m
diff --git a/fs/exec.c b/fs/exec.c
index 7d88d7750f89..d4b67ac9cd53 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1898,20 +1898,12 @@ int do_execve_file(struct file *file, void *__argv, void *__envp)
 	return __do_execve_file(AT_FDCWD, NULL, argv, envp, 0, file);
 }
 
-
-#ifdef CONFIG_KSU
-extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
-			                   void *envp, int *flags);
-#endif
 int do_execve(struct filename *filename,
 	const char __user *const __user *__argv,
 	const char __user *const __user *__envp)
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
 	struct user_arg_ptr envp = { .ptr.native = __envp };
-#ifdef CONFIG_KSU
-	ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
-#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
@@ -1939,9 +1931,6 @@ static int compat_do_execve(struct filename *filename,
 		.is_compat = true,
 		.ptr.compat = __envp,
 	};
-#ifdef CONFIG_KSU
-	ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
-#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
diff --git a/fs/open.c b/fs/open.c
index f17530b2b1f0..13d70db567e0 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -451,15 +451,8 @@ long do_faccessat(int dfd, const char __user *filename, int mode)
 	return res;
 }
 
-#ifdef CONFIG_KSU
-extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
-			 int *flags);
-#endif
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
-#ifdef CONFIG_KSU
-	ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
-#endif
 	return do_faccessat(dfd, filename, mode);
 }
 
diff --git a/fs/read_write.c b/fs/read_write.c
index ee9739afa68a..ec2d20ddd766 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -586,17 +586,8 @@ ssize_t ksys_read(unsigned int fd, char __user *buf, size_t count)
 	return ret;
 }
 
-#ifdef CONFIG_KSU
-extern bool ksu_vfs_read_hook __read_mostly;
-extern int ksu_handle_sys_read(unsigned int fd, char __user **buf_ptr,
-			size_t *count_ptr);
-#endif
 SYSCALL_DEFINE3(read, unsigned int, fd, char __user *, buf, size_t, count)
 {
-#ifdef CONFIG_KSU
-	if (unlikely(ksu_vfs_read_hook))
-		ksu_handle_sys_read(fd, &buf, &count);
-#endif
 	return ksys_read(fd, buf, count);
 }
 
diff --git a/fs/stat.c b/fs/stat.c
index d0dd8de3bf7e..376543199b5a 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -353,9 +353,6 @@ SYSCALL_DEFINE2(newlstat, const char __user *, filename,
 	return cp_new_stat(&stat, statbuf);
 }
 
-#ifdef CONFIG_KSU
-extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
-#endif
 #if !defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_SYS_NEWFSTATAT)
 SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 		struct stat __user *, statbuf, int, flag)
@@ -363,9 +360,6 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 	struct kstat stat;
 	int error;
 
-#ifdef CONFIG_KSU
-	ksu_handle_stat(&dfd, &filename, &flag);
-#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
@@ -516,9 +510,6 @@ SYSCALL_DEFINE4(fstatat64, int, dfd, const char __user *, filename,
 	struct kstat stat;
 	int error;
 
-#ifdef CONFIG_KSU
-	ksu_handle_stat(&dfd, &filename, &flag);
-#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
diff --git a/kernel/reboot.c b/kernel/reboot.c
index 3bf9d23720ad..19f89c07460a 100644
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -305,9 +305,6 @@ DEFINE_MUTEX(system_transition_mutex);
  *
  * reboot doesn't sync: do that yourself before calling this.
  */
-#ifdef CONFIG_KSU
-extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
-#endif
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
@@ -315,10 +312,6 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	char buffer[256];
 	int ret = 0;
 
-#ifdef CONFIG_KSU 
-	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
-#endif
-
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 0ca4d7399130..ef872b2d431f 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -91,10 +91,6 @@
 #include <linux/bpf.h>
 #include <linux/kernfs.h>
 #include <linux/stringhash.h>	/* for hashlen_string() */
-#ifdef CONFIG_KSU
-#include <asm/memory.h>
-#endif
-
 #include "avc.h"
 #include "objsec.h"
 #include "netif.h"
@@ -332,20 +328,11 @@ static void inode_free_security(struct inode *inode)
 	 * concurrent list_add(), but for better safety against future changes
 	 * in the code, we use list_empty_careful() here.
 	 */
-#ifdef CONFIG_KSU
-    spin_lock(&sbsec->isec_lock);
-    if (isec->on_head) {
-        list_del_init(untagged_addr(&isec->list));
-        isec->on_head = false;
-    }
-    spin_unlock(&sbsec->isec_lock);
-#else
 	if (!list_empty_careful(&isec->list)) {
 		spin_lock(&sbsec->isec_lock);
 		list_del_init(&isec->list);
 		spin_unlock(&sbsec->isec_lock);
 	}
-#endif
 }
 
 static void superblock_free_security(struct super_block *sb)
@@ -551,12 +538,7 @@ static int sb_finish_set_opts(struct super_block *sb)
 				list_first_entry(&sbsec->isec_head,
 					   struct inode_security_struct, list);
 		struct inode *inode = isec->inode;
-#ifdef CONFIG_KSU
-        list_del_init(untagged_addr(&isec->list));
-        isec->on_head = false;
-#else
 		list_del_init(&isec->list);
-#endif
 		spin_unlock(&sbsec->isec_lock);
 		inode = igrab(inode);
 		if (inode) {
@@ -1411,9 +1393,6 @@ static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dent
 		spin_lock(&sbsec->isec_lock);
 		if (list_empty(&isec->list)) {
 			list_add(&isec->list, &sbsec->isec_head);
-#ifdef CONFIG_KSU
-			isec->on_head = true;
-#endif
 		}
 		spin_unlock(&sbsec->isec_lock);
 		goto out_unlock;
diff --git a/security/selinux/include/objsec.h b/security/selinux/include/objsec.h
index 49996da38b23..627d75e402f2 100644
--- a/security/selinux/include/objsec.h
+++ b/security/selinux/include/objsec.h
@@ -54,9 +54,6 @@ struct inode_security_struct {
 	u16 sclass;		/* security class of this object */
 	unsigned char initialized;	/* initialization flag */
 	spinlock_t lock;
-#ifdef CONFIG_KSU
-	bool on_head;
-#endif
 };
 
 struct file_security_struct {
